pkg/scheduler/framework/plugins/podtopologyspread/filtering.go

```diff
@@ -223,31 +225,42 @@ func (pl *PodTopologySpread) calPreFilterState(pod *v1.Pod) (*preFilterState, er
 		TpPairToMatchNum:     make(map[topologyPair]*int32, sizeHeuristic(len(allNodes), constraints)),
 	}
 	requiredSchedulingTerm := nodeaffinity.GetRequiredNodeAffinity(pod)
+
+	var availableNodes []*framework.NodeInfo
 	for _, n := range allNodes {
 		node := n.Node()
 		if node == nil {
 			klog.ErrorS(nil, "Node not found")
 			continue
 		}
+
+		// Filter the node out if it carries a `node.kubernetes.io/unschedulable:NoSchedule` taint and pod doesn't tolerate it
+		if taints.TaintExists(node.Spec.Taints, unschedulableTaint) && !corev1.TolerationsTolerateTaint(pod.Spec.Tolerations, unschedulableTaint) {
+			continue
+		}
+
 		// In accordance to design, if NodeAffinity or NodeSelector is defined,
 		// spreading is applied to nodes that pass those filters.
 		// Ignore parsing errors for backwards compatibility.
 		match, _ := requiredSchedulingTerm.Match(node)
 		if !match {
 			continue
 		}
+
 		// Ensure current node's labels contains all topologyKeys in 'Constraints'.
 		if !nodeLabelsMatchSpreadConstraints(node.Labels, constraints) {
 			continue
 		}
+
 		for _, c := range constraints {
 			pair := topologyPair{key: c.TopologyKey, value: node.Labels[c.TopologyKey]}
 			s.TpPairToMatchNum[pair] = new(int32)
 		}
+		availableNodes = append(availableNodes, n)
```

Comment:
This seems to reveal and fix a potential bug. Raised #106971 for discussion. If we reach a consensus to think the issue is a bug, I'd like to hold this PR before this pure fix is patched first.

Reply:
yes, I also found this potential problem.